File: csharp/glp_link/seam/LinkAddress.cs
namespace GlpRuntime.Link.Seam;

/// <summary>
/// The host/path (and optional port) a transport leaf opens against. Mirrors the
/// GLP <c>Endpoint ::= String ; ep(String, Integer)</c> component of
/// <c>link_id(Scheme, Endpoint, Nonce)</c>
/// (contracts/link-primitives.md §1, line 55).
/// </summary>
/// <param name="Host">
/// The host or path. For <c>file</c> a filesystem path; for <c>ws</c>/<c>mqtt</c>
/// a host (or full URL path); for <c>loopback</c> a logical channel name.
/// </param>
/// <param name="Port">
/// The port for the <c>ep(Host, Port)</c> form, or <c>null</c> for the bare-string
/// <c>Endpoint ::= String</c> form (path/host with no port).
/// </param>
public sealed record LinkAddress(string Host, int? Port = null)
{
    /// <summary>The bare-string <c>Endpoint ::= String</c> form (path or host, no port).</summary>
    public static LinkAddress Path(string hostOrPath)
    {
        if (string.IsNullOrWhiteSpace(hostOrPath))
            throw new ArgumentException("LinkAddress requires a non-blank host/path.", nameof(hostOrPath));
        return new LinkAddress(hostOrPath, null);
    }

    /// <summary>The <c>ep(Host, Port)</c> form (host + port).</summary>
    public static LinkAddress Endpoint(string host, int port)
    {
        if (string.IsNullOrWhiteSpace(host))
            throw new ArgumentException("LinkAddress requires a non-blank host.", nameof(host));
        if (port is < 0 or > 65535)
            throw new ArgumentOutOfRangeException(nameof(port), port, "Port must be in [0, 65535].");
        return new LinkAddress(host, port);
    }

    public override string ToString() => Port is { } p ? $"{Host}:{p}" : Host;
}

File: glp_runtime/lib/link/seam/link_address.dart
/// The host/path (and optional port) a transport leaf opens against. Mirrors the
/// GLP `Endpoint ::= String ; ep(String, Integer)` component of
/// `link_id(Scheme, Endpoint, Nonce)`
/// (contracts/link-primitives.md §1, line 55).
///
/// [host]: The host or path. For `file` a filesystem path; for `ws`/`mqtt`
/// a host (or full URL path); for `loopback` a logical channel name.
///
/// [port]: The port for the `ep(Host, Port)` form, or `null` for the bare-string
/// `Endpoint ::= String` form (path/host with no port).
class LinkAddress {
  final String host;
  final int? port;

  const LinkAddress(this.host, [this.port]);

  /// The bare-string `Endpoint ::= String` form (path or host, no port).
  static LinkAddress path(String hostOrPath) {
    if (hostOrPath.trim().isEmpty) {
      throw ArgumentError.value(
          hostOrPath, 'hostOrPath', 'LinkAddress requires a non-blank host/path.');
    }
    return LinkAddress(hostOrPath, null);
  }

  /// The `ep(Host, Port)` form (host + port).
  static LinkAddress endpoint(String host, int port) {
    if (host.trim().isEmpty) {
      throw ArgumentError.value(
          host, 'host', 'LinkAddress requires a non-blank host.');
    }
    if (port < 0 || port > 65535) {
      throw RangeError.range(port, 0, 65535, 'port', 'Port must be in [0, 65535].');
    }
    return LinkAddress(host, port);
  }

  @override
  bool operator ==(Object other) =>
      other is LinkAddress && host == other.host && port == other.port;

  @override
  int get hashCode => Object.hash(host, port);

  @override
  String toString() => port != null ? '$host:$port' : host;
}
